### PR TITLE
[LW-10396] derive midnight addresses

### PR DESCRIPTION
**Pull request by VanessaPC**
# Checklist

- [x] JIRA - lw-10369
- [ ] Proper tests implemented
- [x] Screenshots added.

---

## Proposed solution

Set up and configure Midnight package. 
Add support for `createMidnightWallet`.
Save address and encrypted serialised wallet state in the wallet repo as part of the wallet metadata as `midnightWalletAddress` and `encryptedSerializedMidnightWallet` .

## Testing
- Create new wallet, check that the new address as `midnightWalletAddress` is saved to the wallet metadata in IndexDB. 

## Screenshots

Screenshot 2024-05-28 at 09 19 10

### CODE REVIEW

File: apps/browser-extension-wallet/webpack.common.app.js [hunk 1]
```diff
@@ -20,7 +20,8 @@ module.exports = () =>
       ['trezor-usb-permissions']: path.join(__dirname, 'src/lib/scripts/trezor/trezor-usb-permissions.ts')
     },
     experiments: {
-      syncWebAssembly: true
+      syncWebAssembly: true,
+      topLevelAwait: true
```
Comment (lucas-barros) [inline apps/browser-extension-wallet/webpack.common.app.js:24]: Where was `topLevelAwait` needed?

File: packages/midnight/package.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    "prepack": "yarn build",
+    "prepare": "ts-patch install -s",
+    "prestart": "yarn build",
+    "test:coverage": "yarn test --coverage",
+    "type-check": "echo \"@lace/midnight: no type-check command specified\"",
+    "watch": "yarn build --watch"
+  },
+  "dependencies": {
+    "@cardano-sdk/key-management": "0.15.1",
+    "@midnight-ntwrk/dapp-connector-api": "^1.1.0",
```
Comment (lucas-barros) [inline packages/midnight/package.json:42]: is `@midnight-ntwrk/dapp-connector-api` for future use? I could find any usage.

File: packages/midnight/package.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
+{
+  "name": "@lace/midnight",
+  "version": "0.1.0",
+  "private": true,
+  "homepage": "https://github.com/input-output-hk/wallet-world/blob/master/packages/midnight/README.md",
```
Comment (lucas-barros) [inline packages/midnight/package.json:5]:
```suggestion
  "homepage": "https://github.com/input-output-hk/lace/blob/master/packages/midnight/README.md",
```

File: packages/midnight/package.json [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  "bugs": {
+    "url": "https://github.com/input-output-hk/wallet-world/issues"
```
Comment (lucas-barros) [inline packages/midnight/package.json:7]:
```suggestion
    "url": "https://github.com/input-output-hk/lace/issues"
```

File: packages/midnight/package.json [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/input-output-hk/wallet-world.git"
```
Comment (lucas-barros) [inline packages/midnight/package.json:11]:
```suggestion
    "url": "git+https://github.com/input-output-hk/lace.git"
```

File: packages/midnight/package.json [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  "files": [
+    "dist",
+    "LICENSE",
+    "NOTICE",
+    "README.md"
+  ],
+  "scripts": {
+    "build": "npx rollup -c rollup.config.js",
+    "cleanup": "shx rm -rf dist node_modules coverage reports .rollup.cache src/dist",
+    "format-check": "echo \"@lace/common: no format-check command specified\"",
```
Comment (lucas-barros) [inline packages/midnight/package.json:31]:
```suggestion
    "format-check": "echo \"@lace/midnight: no format-check command specified\"",
```

File: packages/midnight/package.json [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    "@midnight-ntwrk/dapp-connector-api": "^1.1.0",
+    "@midnight-ntwrk/wallet": "^3.5.9",
+    "@midnight-ntwrk/wallet-api": "^3.3.1",
+    "@midnight-ntwrk/zswap": "^0.3.11"
+  },
+  "devDependencies": {
+    "shx": "^0.3.2",
+    "typescript": "^4.3.5"
+  },
+  "peerDependencies": {
```
Comment (lucas-barros) [inline packages/midnight/package.json:51]: I didnt see any packages from `peerDependencies` been used, maybe we don't need them here?

File: packages/midnight/rollup.config.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import commonjs from '@rollup/plugin-commonjs';
+import typescript from '@rollup/plugin-typescript';
+import postcss from 'rollup-plugin-postcss';
+import svgr from '@svgr/rollup';
+import copy from 'rollup-plugin-copy';
+import json from '@rollup/plugin-json';
+
+import packageJson from './package.json';
+
+const baseConfig = (tsConfigPath) => ({
```
Comment (lucas-barros) [inline packages/midnight/rollup.config.js:13]: Could the `baseConfing` from root` rollup.config.js` be used here?

File: packages/midnight/src/tsconfig.background.json [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "rootDir": "./",
+    "outDir": "../dist",
+    "declarationDir": "../dist"
+  },
+  "include": ["./background", "./config", "../lib"],
```
Comment (lucas-barros) [inline packages/midnight/src/tsconfig.background.json:8]: I dont see`background` and `lib` folders, are they needed here?

File: packages/midnight/rollup.config.js (hunk 8 above)
Comment (VanessaPC) [reply packages/midnight/rollup.config.js:13]: Rollup has configuration set up different to `baseConfig` that helps deal with their `wallet-api` which has issues with Scala code conversions. So we need it.

File: apps/browser-extension-wallet/webpack.common.app.js (hunk 1 above)
Comment (VanessaPC) [reply apps/browser-extension-wallet/webpack.common.app.js:24]: This is used by the midnight package. In order to be able to be used in the extension code.

File: packages/midnight/package.json [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    "lint": "cd ../.. && yarn midnight:lint",
+    "prepack": "yarn build",
+    "prepare": "ts-patch install -s",
+    "prestart": "yarn build",
+    "test:coverage": "yarn test --coverage",
+    "type-check": "echo \"@lace/midnight: no type-check command specified\"",
+    "watch": "yarn build --watch"
+  },
+  "dependencies": {
+    "@cardano-sdk/key-management": "0.15.1",
```
Comment (pczeglik-iohk) [inline packages/midnight/package.json:41]: Can you please check if key-management `0.20.5` works with midnight. This is the version we use in `packages/cardano`

File: commitlint.config.js [hunk 11]
```diff
@@ -34,7 +34,8 @@ module.exports = {
         'ui',
         // SRE
         'monitoring',
-        'security'
+        'security',
+        'midnight'
```
Comment (pczeglik-iohk) [inline commitlint.config.js:38]: can you move this up, between `extension` and `staking`? In this current place it suggests that `midnight` is **SRE** related scope

Comment (pczeglik-iohk) [reply commitlint.config.js:38]: probably `security` is also in wrong place ;)

File: packages/midnight/src/tsconfig.background.json (hunk 9 above)
Comment (VanessaPC) [reply packages/midnight/src/tsconfig.background.json]: This has been addressed here 0ffe5f172f8eb0f90c7642380fa572e118b1b27c

File: packages/midnight/package.json (hunk 7 above)
Comment (VanessaPC) [reply packages/midnight/package.json]: This has been addressed here 0ffe5f172f8eb0f90c7642380fa572e118b1b27c

File: packages/midnight/package.json (hunk 6 above)
Comment (VanessaPC) [reply packages/midnight/package.json]: This has been addressed here 0ffe5f172f8eb0f90c7642380fa572e118b1b27c

File: packages/midnight/package.json (hunk 5 above)
Comment (VanessaPC) [reply packages/midnight/package.json]: This has been addressed here 0ffe5f172f8eb0f90c7642380fa572e118b1b27c

File: packages/midnight/package.json (hunk 4 above)
Comment (VanessaPC) [reply packages/midnight/package.json]: This has been addressed here 0ffe5f172f8eb0f90c7642380fa572e118b1b27c

File: packages/midnight/package.json (hunk 3 above)
Comment (VanessaPC) [reply packages/midnight/package.json]: This has been addressed here 0ffe5f172f8eb0f90c7642380fa572e118b1b27c

File: packages/midnight/package.json (hunk 2 above)
Comment (VanessaPC) [reply packages/midnight/package.json]: This has been addressed here 0ffe5f172f8eb0f90c7642380fa572e118b1b27c

File: packages/midnight/package.json (hunk 10 above)
Comment (VanessaPC) [reply packages/midnight/package.json]: Thanks, it works, addressed here: 40a64be9ec5dfab39eacbe32c28ca633b43f7e41

File: commitlint.config.js (hunk 11 above)
Comment (VanessaPC) [reply commitlint.config.js]: True, addressed here: 40a64be9ec5dfab39eacbe32c28ca633b43f7e41
Additionally, Security didn't seem to fee within SRE as you mentioned, but neither within the packages, so I added its own commented off title as `//Security` which seems more fitting. Let me know if there is another preferred way.

File: apps/browser-extension-wallet/src/hooks/__tests__/useWalletManager.test.tsx [hunk 12]
```diff
@@ -334,7 +334,7 @@ describe('Testing useWalletManager hook', () => {
   });
 
   describe('createWallet', () => {
-    test('should store wallet in repository', async () => {
+    test.only('should store wallet in repository', async () => {
```
Comment (lucas-barros) [inline apps/browser-extension-wallet/src/hooks/__tests__/useWalletManager.test.tsx:337]: Left over `only`

Comment (VanessaPC) [reply apps/browser-extension-wallet/src/hooks/__tests__/useWalletManager.test.tsx]: fixed here: 5b99c9603f0eaeb915b62fb8ed36ca6999b6b067

**Review by lucas-barros — APPROVED** [review]

**Review by pczeglik-iohk — APPROVED** [review]

**Pull request closed without merging**